### PR TITLE
[FrameworkBundle][Translation] Add support for Translator paths, Twig paths and Translator aware services paths in commands

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29085, #29633, #17739
| License       | MIT
| Doc PR        | TODO

Add custom (also common) Twig and Translation paths to the translation commands:
 * Custom directories configured in `twig.paths`.
 * Custom directories configured in `translator.paths`
 * The `Resources/translations/` directory of `Validation` component (if installed).
 * The `Resources/translations/` directory of `Form` component (if installed).
 * The `Resources/translations/` directory of Security Core component (if installed).

### CODE REVIEW

**Review by nicolas-grekas — COMMENTED** [review]
thanks for starting this :)

File: src/Symfony/Component/Translation/DependencyInjection/TranslatorPass.php [hunk 1]
```diff
@@ -68,12 +68,16 @@ public function process(ContainerBuilder $container)
...
+        $paths = $container->getDefinition('twig.template_iterator')->getArgument(2);
+
         if ($container->hasDefinition($this->debugCommandServiceId)) {
             $container->getDefinition($this->debugCommandServiceId)->replaceArgument(4, $container->getParameter('twig.default_path'));
+            $container->getDefinition($this->debugCommandServiceId)->replaceArgument(5, $paths);
         }
 
         if ($container->hasDefinition($this->updateCommandServiceId)) {
             $container->getDefinition($this->updateCommandServiceId)->replaceArgument(5, $container->getParameter('twig.default_path'));
+            $container->getDefinition($this->updateCommandServiceId)->replaceArgument(7, $paths);
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Component/Translation/DependencyInjection/TranslatorPass.php]: 7 for `twigPaths` ;)

**Comment by yceruto** [comment]
**Update:** Added extracting translation messages from services that have the translator injected into them by using the following service injection methods: (properties, bindings, construct arguments and method calls).

BUT they're other methods that I can't to reach (e.g. the service locator related to: services subscriber and constroller service arguments) I need help to accomplish it, maybe @nicolas-grekas ?

**Review by nicolas-grekas — COMMENTED** [review]
we should maybe split the translator propagation in another pass.
I'd suggest to look at the HotPathPass also: we could looks at services that has the translation injected *transitively*. And that would give you a hint about following locators btw (HotPathPass doesn't follow them but here we would - see `instanceof ArgumentInterface` + its getValue method).

**Comment by yceruto** [comment]
Status: Needs work

**Comment by yceruto** [comment]
Ready for another round :)

Status: Needs Review

**Review by damienalexandre — APPROVED** [review]
As stated in https://github.com/symfony/symfony/issues/29633, what I like about this is that the discovery of translation files from FrameworkBundle is now also used in the debug:translation command :+1:

File: src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php [hunk 2]
```diff
@@ -99,7 +100,8 @@ public function build(ContainerBuilder $container)
         $container->addCompilerPass(new AddAnnotationsCachedReaderPass(), PassConfig::TYPE_AFTER_REMOVING, -255);
         $this->addCompilerPassIfExists($container, AddValidatorInitializersPass::class);
         $this->addCompilerPassIfExists($container, AddConsoleCommandPass::class, PassConfig::TYPE_BEFORE_REMOVING);
-        $this->addCompilerPassIfExists($container, TranslatorPass::class);
+        $this->addCompilerPassIfExists($container, TranslatorPass::class, PassConfig::TYPE_BEFORE_OPTIMIZATION, -32);
```
[the remark this answers is not in the document]
Comment (fabpot) [reply src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php]: I think adding a comment here would be useful (to remember why we have this priority)

**Comment by yceruto** [comment]
Should I solve here all fabbot issues related to use short array?

**Comment by nicolas-grekas** [comment]
yes, please fix fabbot issues or your PR will conflict soon.

**Comment by yceruto** [comment]
Hellou! project demo ready https://github.com/yceruto/new_trans_paths (read README for instructions) and see diff here https://github.com/yceruto/new_trans_paths/commit/60d7f8cfb59e207e077052a9cdade7d9e1741973, 

The new feature is increasing the reliability of the `debug:translation` and `translation:update` commands, check it out!

**Comment by yceruto** [comment]
Yay! funcional tests passed!

@nicolas-grekas this is ready on my side, how does it look for you?

**Comment by yceruto** [comment]
Friendly ping.

**Comment by yceruto** [comment]
Maybe I should split this PR into two for an easier review? I mean, move the last point that contains a little more complexity:
> * All PHP files from services "`translator` aware" via constructor argument, public property, method calls, service subscriber, controller argument)

**Comment by xabbuh** [comment]
@yceruto I think that could be helpful

**Comment by yceruto** [comment]
@xabbuh working on it, thanks for your reply.

**Comment by yceruto** [comment]
**Update:** I've divided the complexity of this PR as I mentioned earlier, see the updated description.

Status: Needs review

**Comment by yceruto** [comment]
Done https://github.com/symfony/symfony/pull/30120.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @yceruto.

**Pull request merged**